src/csharp/Grpc.Core/Interceptors/ClientInterceptor.cs

```diff
@@ -0,0 +1,281 @@
+            Func<Func<Metadata>, Func<Metadata>> getTrailers = null,
+            Func<Action, Action> dispose = null)
+        {
+            GrpcPreconditions.CheckNotNull(call, "call");
+            var callResponse = call.ResponseAsync;
+            Task<Metadata> callResponseHeaders = call.ResponseHeadersAsync;
+            Func<Status> callStatus = call.GetStatus;
+            Func<Metadata> callTrailers = call.GetTrailers;
+            Action callDispose = call.Dispose;
+
+            if (callResponse != null)
+            {
+                var responseOriginal = callResponse;
+                callResponse = new Task<TResponse>(() => response(responseOriginal));
+            }
+
+            if (responseHeaders != null)
+            {
+                var callResponseHeadersOriginal = callResponseHeaders; // Ensure it will not be captured by the closure
+                callResponseHeaders = new Task<Metadata>(() => responseHeaders(callResponseHeadersOriginal));
+            }
+
+            if (getStatus != null)
+            {
+                callStatus = getStatus(callStatus);
+            }
+
+            if (getTrailers != null)
+            {
+                callTrailers = getTrailers(callTrailers);
+            }
+
+            if (dispose != null)
+            {
+                callDispose = dispose(callDispose);
+            }
+
+            return new AsyncUnaryCall<TResponse>(callResponse, callResponseHeaders, callStatus, callTrailers, callDispose);
+        }
+
+        /// <summary>
+        /// Returns a new instance of <c>AsyncServerStreamingCall</c> that has hooks to intercept on an underlying object of the same type.
+        /// </summary>
+        protected static AsyncServerStreamingCall<TResponse> Intercept<TRequest, TResponse>(AsyncServerStreamingCall<TResponse> call,
+            Func<IAsyncStreamReader<TResponse>, IAsyncStreamReader<TResponse>> responseStream = null,
+            Func<Task<Metadata>, Metadata> responseHeaders = null,
+            Func<Func<Status>, Func<Status>> getStatus = null,
+            Func<Func<Metadata>, Func<Metadata>> getTrailers = null,
+            Func<Action, Action> dispose = null)
+        {
+            GrpcPreconditions.CheckNotNull(call, "call");
+            var callResponseStream = call.ResponseStream;
+            Task<Metadata> callResponseHeaders = call.ResponseHeadersAsync;
+            Func<Status> callStatus = call.GetStatus;
+            Func<Metadata> callTrailers = call.GetTrailers;
+            Action callDispose = call.Dispose;
+
+            if (responseStream != null)
+            {
+                callResponseStream = responseStream(callResponseStream);
+            }
+
+            if (responseHeaders != null)
+            {
+                var callResponseHeadersOriginal = callResponseHeaders; // Ensure it will not be captured by the closure
+                callResponseHeaders = new Task<Metadata>(() => responseHeaders(callResponseHeadersOriginal));
```

Comment:
I have thought about this a bit more and I think there is no good reason for this set of functions to exist, whose purpose is to circumvent the `internal`-ness of the constructors of `Async***Call` objects. If we open up the constructors, this whole thing becomes unnecessary.

Reply:
Downsides:
we'd need to polish the Async**Call constructor signature and be certain that it's not going to change (that would be an API breakage and we are trying hard to avoid that). At the same we would lose some ability to add new members to the Async**Call (what if you need to add more arguments to the constructor later?).
Also, more generally, there is always some risk associated with exposing lower-level APIs to end users, because non-expert users tend to shoot themselves in the foot if you expose too much of the internals - and gRPC users shouldn't be required to become experts before they can use gRPC efficiently.

That said, I think it makes sense to revisit making Async**Call constructors public - now that we have an official support for interceptors coming.

FTR, here's an external PR for publishing the Async**Call constructor - but I think a bit more refactoring would be useful https://github.com/grpc/grpc/pull/9330/files.

Reply:
> we'd need to polish the `Async*Call` constructor signature and be certain that it's not going to change

The members are already public, so we cannot remove them and the constructor at the moment is basically initializing the existing members. We will be able to add additional constructors should we add new members (as long as we can find sensible defaults for values). In effect, by exposing this functionality within this PR under the interceptor API, we are transitively doing most of this API retention cost (admittedly it is a bit contained, but still would be breaking). I think we should seriously consider opening them up, since the class itself and its members are already not `internal` and the `InterceptAsync*` methods are really ugly in this PR.

I do agree that polishing the constructor now makes sense, but I don't see anything wrong with it. What would be the ideal in your opinion?

Reply:
Yes, let's look into publishing the Async**Call constructors in a separate PR (and count on them being already public in this PR).

Reply:
What do you think about not opening the constructors but adding a static factory method that returns a new instance? The advantage would be that we will retain the freedom to *unseal* the class and return a derived instance without an API breakage, but it is debatable that it would be worth much.

Reply:
The PR for publicizing constructors is here: https://github.com/grpc/grpc/pull/12877 though I am still open to the static factory route idea.